### PR TITLE
router: Imrpove router matching by caching sanitized_path in stream_info

### PR DESCRIPTION
**Pull request by bryanwux**
Remove repeative work on sanitizing request headers when routing by caching `sanitized_path` in stream_info.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #21615
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/router/config_impl.cc [hunk 1]
```diff
@@ -1404,10 +1404,11 @@ absl::optional<std::string> PathTemplateRouteEntryImpl::currentUrlPathAfterRewri
...
 }
 
-RouteConstSharedPtr PathTemplateRouteEntryImpl::matches(const Http::RequestHeaderMap& headers,
-                                                        const StreamInfo::StreamInfo& stream_info,
-                                                        uint64_t random_value) const {
+RouteConstSharedPtr PathTemplateRouteEntryImpl::matches(
+    const Http::RequestHeaderMap& headers, const StreamInfo::StreamInfo& stream_info,
+    uint64_t random_value, bool ignore_path_parameters_in_path_matching) const {
   throw absl::UnimplementedError("Path template matcher not implemented");
+  (void)ignore_path_parameters_in_path_matching;
```
Comment (daixiang0) [inline source/common/router/config_impl.cc:1411]: void a bool value?

Comment (bryanwux) [reply source/common/router/config_impl.cc:1411]: void silences compiler unused warning. I will comment on it.

File: source/common/stream_info/stream_info_impl.h [hunk 2]
```diff
@@ -312,6 +314,16 @@ struct StreamInfoImpl : public StreamInfo {
     ASSERT(downstream_bytes_meter_.get() == downstream_bytes_meter.get());
   }
 
+  void setSanitizedPath(const Http::RequestHeaderMap& headers) override {
+    sanitized_path_ = Http::PathUtil::removeQueryAndFragment(headers.getPathValue());
+    auto pos = sanitized_path_.find_first_of(";");
+    if (pos != absl::string_view::npos) {
+      sanitized_path_.remove_suffix(sanitized_path_.length() - pos);
+    }
```
Comment (giantcroc) [inline source/common/stream_info/stream_info_impl.h:322]: Hi, why we need to remove suffix after `;` directly?

File: source/common/router/config_impl.cc (hunk 1 above)
Comment (daixiang0) [reply source/common/router/config_impl.cc:1411]: What about using an anonymous parameter?

File: source/common/stream_info/stream_info_impl.h (hunk 2 above)
Comment (bryanwux) [reply source/common/stream_info/stream_info_impl.h:322]: If `ignore_port_in_host_matching` is set to true, the suffix after `;` will be removed. Like `"/path-bluh;env=prod"` will be sanitized to `"/path-bluh"`.

File: source/common/router/config_impl.cc (hunk 1 above)
Comment (bryanwux) [reply source/common/router/config_impl.cc:1411]: Probably an UNREFERENCED_PARAMETER wrapper on the variable?  WDYT? https://github.com/envoyproxy/envoy/blob/0b10b055a6e1dbdd600f9a80f8b8d803bdd8ec9c/source/common/common/macros.h#L21-L26

Comment (zhxie) [reply source/common/router/config_impl.cc:1411]: You can just simply omit the unused variable name in the function signature to suppress the warning/error.

Comment (bryanwux) [reply source/common/router/config_impl.cc:1411]: Sounds good.

**Comment by zhxie** [comment]
I guess the patch only caches sanitized path (the path without path-parameters), but the performance concern mentioned in the issue is about `removeQueryAndFragment`?

**Comment by bryanwux** [comment]
> I guess the patch only caches sanitized path (the path without path-parameters), but the performance concern mentioned in the issue is about `removeQueryAndFragment`?

True. I will cache the result of `removeQueryAndFragment` beforehand.

File: source/common/stream_info/stream_info_impl.h (hunk 2 above)
Comment (giantcroc) [reply source/common/stream_info/stream_info_impl.h:322]: OK, got it, thanks!

File: envoy/stream_info/stream_info.h [hunk 3]
```diff
@@ -732,6 +732,10 @@ class StreamInfo {
    */
   virtual void setDownstreamBytesMeter(const BytesMeterSharedPtr& downstream_bytes_meter) PURE;
 
+  virtual void setPathWithoutQueryAndFragment(const Http::RequestHeaderMap& headers) PURE;
```
Comment (adisuissa) [inline envoy/stream_info/stream_info.h:735]: Please add comments for the interface.

A different question, why not set the entire path in the stream info, and allow fetching the path without the query and fragment parts?

**Review by adisuissa — COMMENTED** [review]
Left a high-level comment.

Assigning to Yan for further review as it is related to router.
/assign @yanavlasov

File: envoy/stream_info/stream_info.h (hunk 3 above)
Comment (bryanwux) [reply envoy/stream_info/stream_info.h:735]: It's actually the result of `removeQueryAndFragment` will affect the performance. I think set the entire path in the stream info is unnecessary?

File: source/common/stream_info/stream_info_impl.h (hunk 2 above)
Comment (giantcroc) [reply source/common/stream_info/stream_info_impl.h]: > If `ignore_port_in_host_matching` in route configuration is set to true, the suffix after `;` will be removed. Like `"/path-bluh;env=prod"` will be sanitized to `"/path-bluh"`.

OK, got it. Thanks!

**Comment by yanavlasov** [comment]
I think the problem with this change is that if a HTTP filter changes the `:path` header the router will not see this change and will still route based on the original `:path`.

Was there a performance issue found with the finding the start of query in the existing code? I understand that theoretically it can be expensive if the `:path` is long. If this needs to be addressed the only viable way is to add helper methods to the RequestHeaderMap to store string_view of the `:path` without query that can be updated any time `:path` header is changed.

**Comment by bryanwux** [comment]
> Was there a performance issue found with the finding the start of query in the existing code? I understand that theoretically it can be expensive if the `:path` is long. If this needs to be addressed the only viable way is to add helper methods to the RequestHeaderMap to store string_view of the `:path` without query that can be updated any time `:path` header is changed.

Yes, caching the result of `removeQueryAndFragment` will end up with a 5x performance improvement by benchmark test. The problem is there is per-route repeative work here: 
https://github.com/envoyproxy/envoy/blob/f6f56150b119e58f671401fae4cf31c44f93bec0/source/common/common/matchers.cc#L133-L135

**Comment by yanavlasov** [comment]
Are you talking about 5x improvement for a microbenchmark, or a full Envoy system benchmark? I'd expect this change to make almost no difference when measuring request throughput of a complete system.

I do not think this can be done with the `HEADER_FUNC`. This will need to be custom code that overrides the setPath method and adds additional method to access query.

**Pull request closed without merging**